src/lib.rs

```diff
@@ -285,42 +318,78 @@ where
+impl ClamdClient {
+    async fn connect(
+        &mut self,
+    ) -> Result<std::sync::MutexGuard<'_, Framed<SocketWrapper, ClamdZeroDelimitedCodec>>> {
         let codec = ClamdZeroDelimitedCodec::new();
-        match &self.connection_type {
+        self.socket = match &self.connection_type {
             ConnectionType::Oneshot => match &self.socket_type {
-                SocketType::Tcp(address) => Ok(Framed::new(
+                SocketType::Tcp(address) => Some(Arc::new(Mutex::new(Framed::new(
                     SocketWrapper::Tcp(
                         TcpStream::connect(address)
                             .await
                             .map_err(ClamdError::ConnectError)?,
                     ),
                     codec,
-                )),
-                SocketType::Unix(path) => Ok(Framed::new(
+                )))),
+                SocketType::Unix(path) => Some(Arc::new(Mutex::new(Framed::new(
                     SocketWrapper::Unix(
                         UnixStream::connect(path)
                             .await
                             .map_err(ClamdError::ConnectError)?,
                     ),
                     codec,
-                )),
+                )))),
             },
-            ConnectionType::KeepAlive => todo!(),
-        }
+            ConnectionType::KeepAlive => {
+                if self.socket.is_none() {
+                    match &self.socket_type {
+                        SocketType::Tcp(address) => {
+                            let stream = TcpStream::connect(address).await?;
+                            let socket_ref = SockRef::from(&stream);
+                            socket_ref.set_keepalive(true)?;
+                            let socket = Arc::new(Mutex::new(Framed::new(
+                                SocketWrapper::Tcp(stream),
+                                codec,
+                            )));
+                            let tmp = socket.clone();
+                            let mut sock = tmp.lock().unwrap();
+                            sock.send(ClamdRequestMessage::StartSession).await?;
+                            Some(socket)
+                        }
+                        SocketType::Unix(path) => {
+                            let stream = UnixStream::connect(path)
+                                .await
+                                .map_err(ClamdError::ConnectError)?;
+                            let socket = Arc::new(Mutex::new(Framed::new(
+                                SocketWrapper::Unix(stream),
+                                codec,
+                            )));
+                            let tmp = socket.clone();
+                            let mut sock = tmp.lock().unwrap();
+                            sock.send(ClamdRequestMessage::StartSession).await?;
+                            Some(socket)
+                        }
+                    }
+                } else {
+                    self.socket.clone()
```

Comment:
Why do you do a clone here? This re-clones `self.socket`, maybe just set `self.socket` in the if branch and drop the `else` branch completely.